src/blockmap.jl

```diff
@@ -0,0 +1,297 @@
+
+    for A in As
+        if !(A isa UniformScaling)
+            eltype(A) == T || throw(ArgumentError("eltype type mismatch in vcat of linear maps"))
+        end
+    end
+
+    ncols = 0
+    # find first non-UniformScaling to detect number of columns
+    for A in As
+        if !(A isa UniformScaling)
+            ncols = size(A, 2)
+            break
+        end
+    end
+    ncols == 0 && throw(ArgumentError("hcat of only UniformScaling-like objects cannot determine the linear map size"))
+
+    maps = promote_to_lmaps(ntuple(i->ncols, nbr), 1, T, As...)
+    check_dims(maps, 2)
+    return BlockMap(maps, ntuple(i->1, length(As)))
+end
+
+############
+# hvcat
+############
+
+function Base.hvcat(rows::NTuple{nr,Int}, As::Union{LinearMap,UniformScaling}...) where nr
+    T = promote_type(map(eltype, As)...)
+    sum(rows) == length(As) || throw(ArgumentError("mismatch between row sizes and number of arguments"))
+    n = fill(-1, length(As))
+    needcols = false # whether we also need to infer some sizes from the column count
+    j = 0
+    for i in 1:nr # infer UniformScaling sizes from row counts, if possible:
+        ni = -1 # number of rows in this block-row, -1 indicates unknown
+        for k in 1:rows[i]
+            if !isa(As[j+k], UniformScaling)
+                na = size(As[j+k], 1)
+                ni >= 0 && ni != na &&
+                    throw(DimensionMismatch("mismatch in number of rows"))
+                ni = na
+            end
+        end
+        if ni >= 0
+            for k = 1:rows[i]
+                n[j+k] = ni
+            end
+        else # row consisted only of UniformScaling objects
+            needcols = true
+        end
+        j += rows[i]
+    end
+    if needcols # some sizes still unknown, try to infer from column count
+        nc = -1
+        j = 0
+        for i in 1:nr
+            nci = 0
+            rows[i] > 0 && n[j+1] == -1 && (j += rows[i]; continue)
+            for k = 1:rows[i]
+                nci += isa(As[j+k], UniformScaling) ? n[j+k] : size(As[j+k], 2)
+            end
+            nc >= 0 && nc != nci && throw(DimensionMismatch("mismatch in number of columns"))
+            nc = nci
+            j += rows[i]
+        end
+        nc == -1 && throw(ArgumentError("sizes of UniformScalings could not be inferred"))
+        j = 0
+        for i in 1:nr
+            if rows[i] > 0 && n[j+1] == -1 # this row consists entirely of UniformScalings
+                nci = nc ÷ rows[i]
+                nci * rows[i] != nc && throw(DimensionMismatch("indivisible UniformScaling sizes"))
+                for k = 1:rows[i]
+                    n[j+k] = nci
+                end
+            end
+            j += rows[i]
+        end
+    end
+
+    return BlockMap(promote_to_lmaps(n, 1, T, As...), rows)
+end
+
+promote_to_lmaps_(n::Int, ::Type{T}, J::UniformScaling) where {T} = UniformScalingMap(convert(T, J.λ), n)
+promote_to_lmaps_(n::Int, ::Type{T}, A::LinearMap{T}) where {T} = A
```

Comment:
This could include a dims check, which of course requires that you know if `n` is the row size or column size.

Reply:
That is probably doable, but tricky, and requires careful inspection of how `n` is populated for blocks that have a well-defined size.